lib/cdo/i18n_string_url_tracker.rb

```diff
@@ -76,18 +76,13 @@ def log(string_key, url, source)
 
   # Records the log data to a buffer which will eventually be flushed
   def add_to_buffer(string_key, url, source)
-    data = {url => {string_key => Set[source]}}
     # make sure this is the only thread modifying @buffer
     @buffer.synchronize do
-      # update the buffer size if we are adding any new data to it
-      # duplicate data will not increase the buffer size
-      size = 0
-      size += url.bytesize unless @buffer.dig(url)
-      size += string_key.bytesize unless @buffer.dig(url, string_key)
-      size += source.bytesize unless @buffer.dig(url, string_key)&.include?(source)
-      @buffer_size += size
       # add the new data to the buffer
-      @buffer.deep_merge!(data)
+      ((@buffer[url] ||= {})[string_key] ||= Set.new).add?(source) &&
```

Comment:
What is `deep_merge!` doing if it isn't modifying the hash directly? The code written here is what I thought `deep_merge!` was doing.

Reply:
Yeah, I thought the same but it appears not quite- ActiveSupport's [`Hash#deep_merge!`](https://github.com/rails/rails/blob/main/activesupport/lib/active_support/core_ext/hash/deep_merge.rb) only modifies the outermost Hash in-place, any inner Hashes are `dup`ed before being merged in for some reason.

So `{a: {b: :c}}.deep_merge! {a: {x: :y}}` will cause the `{b: :c}` Hash-object to be duplicated before merging `{x: :y}` into it.

I took a look through the Rails commit history but didn't find an actual reason behind this behavior, perhaps it's just a bug/oversight that can eventually fixed/optimized.